### ISSUE TITLE
Calico rules not inserted at the top of FORWARD chain

### ISSUE THREAD

**Issue by nmeessen** [projectcalico/felix#2132]
## Expected Behavior
Felix is configured with the default behaviour "ChainInsertMode=Insert" and “IptablesRefreshInterval” is non-zero (enabled). With those settings, I would expect calico to maintain  its rules at the top of the chains.

## Current Behavior
Calico rules are added below docker rules.
If the calico-felix service is restarted, it fixes the issue and calico is reinserted at the top of the FORWARD chain.

## Possible Solution
There should be a check that verifies that calico rules are at the top of the forward chain and moves them there if they're not.

## Steps to Reproduce (for bugs)
This issue is most likely triggered due to a very specific sequence of events when a node boots up. Due to how our environment is, it's hard to make a change to the node as it boots. Nevertheless, I think calico should maintain its rules at the top of the chain as suggested by the doc and this is not happening here.

logged in as early as possible: no iptables rules
```
root@ip-10-117-148-70:/home/ubuntu# iptables -L -n -v
Chain INPUT (policy ACCEPT 8 packets, 634 bytes)
 pkts bytes target     prot opt in     out     source               destination         
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
root@ip-10-117-148-70:/home/ubuntu# iptables -L -n -v
Chain INPUT (policy ACCEPT 166 packets, 49497 bytes)
 pkts bytes target     prot opt in     out     source               destination         
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
Chain OUTPUT (policy ACCEPT 151 packets, 36948 bytes)
 pkts bytes target     prot opt in     out     source               destination
```         
a few second later, docker starts up, the docker rules show up
```
root@ip-10-117-148-70:/home/ubuntu# iptables -L -n -v
Chain INPUT (policy ACCEPT 128 packets, 23759 bytes)
 pkts bytes target     prot opt in     out     source               destination         
Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DOCKER-USER  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
    0     0 DOCKER-ISOLATION-STAGE-1  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0           
Chain OUTPUT (policy ACCEPT 135 packets, 22403 bytes)
 pkts bytes target     prot opt in     out     source               destination         
Chain DOCKER (1 references)
 pkts bytes target     prot opt in     out     source               destination         
Chain DOCKER-ISOLATION-STAGE-1 (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DOCKER-ISOLATION-STAGE-2  all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0           
Chain DOCKER-ISOLATION-STAGE-2 (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DROP       all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0           
Chain DOCKER-USER (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0           
root@ip-10-117-148-70:/home/ubuntu# 
```
Felix is actually trying to come up but temporarily failed to resolve the etcd server name (it eventually succeeds)
```
root@ip-10-117-148-70:/home/ubuntu# service calico-felix status
● calico-felix.service - Calico Felix agent
   Loaded: loaded (/lib/systemd/system/calico-felix.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2019-08-27 05:01:31 UTC; 22s ago
 Main PID: 2387 (calico-felix)
    Tasks: 8 (limit: 4633)
   CGroup: /system.slice/calico-felix.service
           └─2387 /usr/bin/calico-felix
Aug 27 05:01:31 ip-10-117-148-70 systemd[1]: Starting Calico Felix agent...
Aug 27 05:01:31 ip-10-117-148-70 systemd[1]: Started Calico Felix agent.
Aug 27 05:01:43 ip-10-117-148-70 calico-felix[2387]: 2019-08-27 05:01:43.149 [ERROR][2387] daemon.go 213: Failed to create datastore client error=dial tcp: lookup etcd01-dev-apse2.foo on 127.0.0.53:53: no such host
```
Felix does its changes to iptables but it appends its rules to the FORWARD chain instead of inserting them at the top???
```
root@ip-10-117-148-70:/home/ubuntu# iptables -L -n -v
Chain INPUT (policy ACCEPT 7 packets, 506 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  286 66041 cali-INPUT  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* cali:Cz_u1IQiXIMmKD4c */
Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DOCKER-USER  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
    0     0 DOCKER-ISOLATION-STAGE-1  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0           
    0     0 cali-FORWARD  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* cali:wUHhoiAYhphO9Mso */
```

## Context
Calico is used in our env to enforce egress restrictions. Traffic from the host hits the OUTPUT chain and is policed accordingly but the traffic from the local (unmanaged) docker which uses the FORWARD chain bypasses calico as it's accepted before hitting the cali-FORWARD rule.

## Your Environment
```
root@ip-10-117-148-70:/home/ubuntu# more /etc/calico/felix.cfg
DatastoreType = etcdv3
EtcdEndpoints = https://etcd01-dev-apse2.foo:2379
EtcdKeyFile = /etc/calico/bar-key.pem
EtcdCertFile = /etc/calico/bar.pem
EtcdCaFile = /etc/calico/bar-ca.pem
FelixHostname = generic-foo
IptablesRefreshInterval = 300
IpsetsRefreshInterval = 240
ReportingIntervalSecs = 0
RouteRefreshInterval = 0
LogSeverityFile = Warning
LogSeverityScreen = Warning
LogSeveritySys = Warning
IptablesFilterAllowAction = RETURN
FailsafeInboundHostPorts = none
FailsafeOutboundHostPorts = none
```
```
root@ip-10-117-148-70:/home/ubuntu# calico-felix --version
Version:            v3.2.0
Full git commit ID: 30458090588d14445f4b16b5b80e6db80dbffc33
Build date:         2018-08-13T07:44:07+0000
```
* Etcd version (if using etcd datastore driver):  3.2.11
* Operating System and version: Ubuntu 18.04.3 LTS

**Comment by fasaxc** [comment]
Please attach a full log from calico-node on a node where you see the issue.

What is your iptables refresh interval set to?  Does it help to reduce it to say 10s or is the refresh running but not having an effect?

**Comment by nmeessen** [comment]
@fasaxc I believe the refresh is running but not having an effect. I created a PR internally today to enable calico debug logs at boot time,  hopefully tomorrow it's deployed and I can generate those logs. As part of that PR, I also changed the IptablesRefreshInterval from 300 to 10 in case it makes a difference.

**Comment by nmeessen** [comment]
[felix-logs.txt](https://github.com/projectcalico/felix/files/3577805/felix-logs.txt)
@fasaxc see the full debug logs attached. Let me know what you can make out of them.

**Comment by nmeessen** [comment]
In case it helps here are the logs from the service restart (after which the cali-FORWARD is inserted at the top of the chain).
[felix-logs-after-service-restart.txt](https://github.com/projectcalico/felix/files/3577850/felix-logs-after-service-restart.txt)

**Comment by rafaelvanoni** [comment]
@fasaxc Can you take a look at the logs?

**Comment by nmeessen** [comment]
Hi @fasaxc did you get a chance to look at the logs? Is there anything I can do to help the troubleshooting?

**Comment by fasaxc** [comment]
Not yet; does this repro on v3.9?  We've done some work in that area since then, there's some chance it's already fixed.

**Comment by fasaxc** [comment]
@nmeessen Your first log finishes after just a couple of seconds, which is not long enough to see a periodic refresh, which is what should sort out the chain if another process writes its rules above ours.

I think we did some work in v3.4 that made us recheck more aggressively, which should help here.  We also reworked the insert machinery in v3.9 (and then backported to a few other releases) to avoid missed updates if both we and docker are trying to add and remove rules.  I'd recommend upgrading to see if either of those fixes this issue.
